api.js

```diff
@@ -640,152 +640,177 @@ function _generateRUMHeader(options = {}, metadata, loader) {
+  )
+  metric.incrementCallCount()
+
+  const { isValidContext, failureIdx, quietMode } = validateContext(this.agent.config)
+
+  if (!isValidContext) {
+    return _gracefail(failureIdx, quietMode)
+  }
+
+  const config = this.agent.config
+
   // This hash gets written directly into the browser.
   const rumHash = {
-    agent: browserMonitoring.js_agent_file,
-    beacon: browserMonitoring.beacon,
-    errorBeacon: browserMonitoring.error_beacon,
-    licenseKey: licenseKey,
-    applicationID: appid,
-    applicationTime: time,
-    transactionName: hashes.obfuscateNameUsingKey(name, key),
-    queueTime: trans.queueTime,
-    ttGuid: trans.id,
+    agent: config.browser_monitoring.js_agent_file,
+    beacon: config.browser_monitoring.beacon,
+    errorBeacon: config.browser_monitoring.error_beacon,
+    licenseKey: config.browser_monitoring.browser_key,
+    applicationID: config.application_id,
 
     // we don't use these parameters yet
     agentToken: null
   }
 
-  const attrs = Object.create(null)
+  const trans = this.agent.getTransaction()
+  const hasActiveTransaction = trans !== null
 
-  const customAttrs = trans.trace.custom.get(ATTR_DEST.BROWSER_EVENT)
-  if (!properties.isEmpty(customAttrs)) {
-    attrs.u = customAttrs
-  }
+  if (hasActiveTransaction) {
+    // bail gracefully outside an ignored transaction
+    if (trans.isIgnored()) {
+      return _gracefail(1)
+    }
 
-  const agentAttrs = trans.trace.attributes.get(ATTR_DEST.BROWSER_EVENT)
-  if (!properties.isEmpty(agentAttrs)) {
-    attrs.a = agentAttrs
-  }
+    /* If we're in an unnamed transaction, add a friendly warning this is to
+     * avoid people going crazy, trying to figure out why browser monitoring is
+     * not working when they're missing a transaction name.
+     */
+    const name = trans.getFullName()
+    if (!name) {
+      return _gracefail(3)
+    }
 
-  if (!properties.isEmpty(attrs)) {
-    rumHash.atts = hashes.obfuscateNameUsingKey(JSON.stringify(attrs), key)
+    const time = trans.timer.getDurationInMillis()
+    rumHash.applicationTime = time
+
+    /*
+     * Only the first 13 chars of the license should be used for hashing with
+     * the transaction name.
+     */
+    const key = config.license_key.substr(0, 13)
+    rumHash.transactionName = hashes.obfuscateNameUsingKey(name, key)
+
+    rumHash.queueTime = trans.queueTime
+    rumHash.ttGuid = trans.id
+
+    const attrs = Object.create(null)
+
+    const customAttrs = trans.trace.custom.get(ATTR_DEST.BROWSER_EVENT)
+    if (!properties.isEmpty(customAttrs)) {
+      attrs.u = customAttrs
+    }
+
+    const agentAttrs = trans.trace.attributes.get(ATTR_DEST.BROWSER_EVENT)
+    if (!properties.isEmpty(agentAttrs)) {
+      attrs.a = agentAttrs
+    }
+
+    if (!properties.isEmpty(attrs)) {
+      rumHash.atts = hashes.obfuscateNameUsingKey(JSON.stringify(attrs), key)
+    }
+  } else {
+    logger.debugOnce(
+      'api:getBrowserTimingHeader',
```

Comment:
nit: would it really be that noisy to use plain `logger.debug` here? I understand server-side rendering is reasonably static and you wouldn't be doing lots of attempts to render without a transaction.

On the other hand, if you accidentally try to inject the header as normal but you're accidentally not in a transaction, getting debug output every time could be helpful.

Reply:
yea this won't matter.  I've noticed in the SSG case it spins up separate threads so the debug once is useless unfortunately